File: hackerrank/easy/sock-merchant.java
// https://www.hackerrank.com/challenges/sock-merchant

import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {
    // Complete the sockMerchant function below.
    static int sockMerchant(int n, int[] ar) {
        Map<Integer, Integer> sockCounts = new HashMap<Integer, Integer>();
        for (int sock : ar) {
            final Integer currentCount = sockCounts.get(sock);
            sockCounts.put(sock, currentCount != null ? currentCount + 1 : 1);
        }

        int pairCount = 0;
        for (Integer sockCount : sockCounts.values()) {
            pairCount += sockCount.intValue() / 2;
        }

        return pairCount;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        int[] ar = new int[n];

        String[] arItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            int arItem = Integer.parseInt(arItems[i]);
            ar[i] = arItem;
        }

        int result = sockMerchant(n, ar);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}

File: hackerrank/easy/sock-merchant.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/sock-merchant

import math
import os
import random
import re
import sys

# Complete the sockMerchant function below.
def sockMerchant(n, ar):
    sockpairs = {}
    for s in ar: sockpairs[s] = sockpairs.get(s, 0) + 1
    numpairs = 0
    for (_, n) in sockpairs.items(): numpairs += n // 2
    return numpairs

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')
    n = int(input())
    ar = list(map(int, input().rstrip().split()))
    result = sockMerchant(n, ar)
    fptr.write(str(result) + '\n')
    fptr.close()
